modules/extensions/src/terrain/terrain-effect.ts

```diff
@@ -0,0 +1,238 @@
+    }
+
+    if (this.heightMap) {
+      this.heightMap.color.delete();
+      this.heightMap.delete();
+      this.heightMap = undefined;
+    }
+
+    for (const terrainCover of this.terrainCovers.values()) {
+      terrainCover.delete();
+    }
+    this.terrainCovers.clear();
+  }
+
+  private _renderHeightMap(
+    gl: WebGLRenderingContext,
+    terrainLayers: Layer[],
+    viewport: Viewport,
+    opts: PreRenderOptions
+  ) {
+    if (!this.heightMap) {
+      // Not supported
+      return;
+    }
+
+    this.heightMap.resize(viewport);
+
+    this.terrainPass.renderHeightMap(this.heightMap, {
+      ...opts,
+      pass: 'terrain-height-map',
+      viewports: [viewport],
+      layers: terrainLayers,
+      moduleParameters: {
+        heightMap: this.heightMap.color,
+        dummyHeightMap: this.dummyHeightMap,
+        devicePixelRatio: 1,
+        drawToTerrainHeightMap: true
+      }
+    });
+  }
+
+  private _updateTerrainCovers(
+    terrainLayers: Layer[],
+    drapeLayers: Layer[],
+    viewport: Viewport,
+    opts: PreRenderOptions
+  ) {
+    if (drapeLayers.length === 0) {
+      // No rerender needed, just disable the texture in the existing terrain covers
+      for (const layer of terrainLayers) {
+        const terrainCover = this.terrainCovers.get(layer.id);
+        terrainCover?.shouldUpdate({layers: drapeLayers});
+      }
+      return;
+    }
+
+    // Mark a terrain cover as dirty if one of the drape layers needs redraw
+    const layerNeedsRedraw: Record<string, boolean> = {};
+    for (const layer of drapeLayers) {
+      if (layer.state.terrainCoverNeedsRedraw) {
+        layerNeedsRedraw[layer.id] = true;
+        layer.state.terrainCoverNeedsRedraw = false;
+      }
+    }
+    for (const terrainCover of this.terrainCovers.values()) {
+      terrainCover.isDirty = terrainCover.isDirty || terrainCover.shouldUpdate({layerNeedsRedraw});
+    }
+
+    for (const layer of terrainLayers) {
+      this._updateTerrainCover(layer, drapeLayers, viewport, opts);
+    }
+
+    if (!this.isPicking) {
+      this._pruneTerrainCovers();
+    }
+  }
+
+  private _updateTerrainCover(
+    terrainLayer: Layer,
+    drapeLayers: Layer[],
+    viewport: Viewport,
+    opts: PreRenderOptions
+  ) {
+    const renderPass = this.isPicking ? this.terrainPickingPass : this.terrainPass;
+    let terrainCover = this.terrainCovers.get(terrainLayer.id);
+    if (!terrainCover) {
+      terrainCover = new TerrainCover(terrainLayer);
+      this.terrainCovers.set(terrainLayer.id, terrainCover);
+    }
+    try {
+      const isDirty = terrainCover.shouldUpdate({
+        targetLayer: terrainLayer,
+        viewport,
+        layers: drapeLayers
+      });
+      if (this.isPicking || terrainCover.isDirty || isDirty) {
+        renderPass.renderTerrainCover(terrainCover, {
+          ...opts,
+          layers: drapeLayers,
+          moduleParameters: {
+            dummyHeightMap: this.dummyHeightMap,
```

Comment:
We are using this `dummyXXXXMap` pattern in a number of effects. Would it be worth having a single global dummy map to avoid having to duplicate resources?

Reply:
@ibgreen Maybe this is a luma question - is it reasonable to add something like `Texture.getDefaultTexture(device)`? These "dummy" textures are never read in the shader, just to make sure that all samplers are supplied for the model to render.